From a1e470704be30a53d1df037988f3442d89f8038a Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Thu, 8 Oct 2026 15:57:46 +0200
Subject: [PATCH] Web documentation of the generic-parallel interfaces

New page generic-parallel.md (loading, integration, models, material
properties, memory, execution policy, errors, limitations, example),
linked from the index, the menu, the C++ introduction, the orthotropic
behaviours and the failure analysis pages. Release notes, installation
options and README updated. The list of the rotation functions of the
finite strain behaviours is fixed.
---
 INSTALL-cmake.md                              |  14 +-
 README.md                                     |  15 +-
 docs/web/CMakeLists.txt                       |   1 +
 .../behaviour-integration-failure-analysis.md |  14 +
 docs/web/bindings-cxx.md                      |  65 +++-
 docs/web/generic-parallel.md                  | 289 ++++++++++++++++++
 docs/web/index.md                             |  17 +-
 docs/web/install.md                           |  12 +-
 docs/web/mgis-template.html                   |   1 +
 docs/web/orthotropic-behaviours.md            |  82 ++++-
 docs/web/release-notes-3.2.md                 |  71 +++++
 11 files changed, 557 insertions(+), 24 deletions(-)
 create mode 100644 docs/web/generic-parallel.md

diff --git a/INSTALL-cmake.md b/INSTALL-cmake.md
index 8273f7e..8fe875c 100644
--- a/INSTALL-cmake.md
+++ b/INSTALL-cmake.md
@@ -27,7 +27,7 @@ The simplest way to compile this package is:
 
 2. Type `make` to compile the package.
 
-4. Optionally, type `make built-tests` to compile all tests come with
+3. Optionally, type `make build-tests` to compile all tests come with
   the package. This is automatically called by `make check`
 
 4. Optionally, type `make check` to run any self-tests that come with
@@ -78,7 +78,17 @@ Options
   not fail if the `HDF5` library is not found. See
   `enable-hdf5-automatic-support` for details.
 - `enable-hdf5-automatic-support`: if set, `MGIS` tries by default to
-  support `HDF5` even if enable-hdf5-support` is not set.
+  support `HDF5` even if `enable-hdf5-support` is not set.
+- `enable-generic-parallel-cuda-tests`: test the behaviours, models and
+  material properties generated by the `generic-parallel` interface of
+  `MFront` with the `cuda` backend. This requires a `CUDA` compiler and an
+  `NVIDIA` GPU (default=OFF).
+
+The tests of the `generic-parallel` interface of `MFront` are only built
+if the `TFEL` libraries found provide this interface. They use the
+`stlpar` backend (parallel algorithms of the `C++` standard library). With
+`gcc`, those algorithms rely on the `TBB` library, which is linked if
+found.
 
 `cmake` usefull variables
 =======================
diff --git a/README.md b/README.md
index 34b3ca6..8ffc3f4 100644
--- a/README.md
+++ b/README.md
@@ -1,12 +1,13 @@
 # MFrontGenericInterfaceSupport
 
 This project aims at providing tools (functions, classes, bindings,
-etc...) to handle behaviours written using `MFront` generic interface.
+etc...) to handle behaviours, models and material properties generated by
+the `generic` and `generic-parallel` interfaces of `MFront`.
 For information about `MFront`, see
 <http://thelfer.github.io/tfel/web/index.html>.
 
 Those tools are meant to be used by (`FEM`, `FFT`, etc.) solver
-developers. This tools are *not* linked to the `TFEL` libraries.
+developers. Those tools are *not* linked to the `TFEL` libraries.
 Permissive licences have been chosen to allow integration in open-source
 and proprietary codes.
 
@@ -32,7 +33,15 @@ library provides two main functions:
   step.
 
 The library also supports handling a group of integration points though
-the `mgis::behaviour::MaterialStateManager` class.
+the `mgis::behaviour::MaterialStateManager` and
+`mgis::behaviour::MaterialDataManager` classes.
+
+The behaviours, models and material properties generated by the
+`generic-parallel` interface of `MFront` process whole batches of
+integration points, on CPUs or on GPUs. The functions acting on a
+`MaterialDataManager` use them automatically. Material properties are
+evaluated on a set of points by the `mgis::material_property::evaluate`
+function. See <https://thelfer.github.io/mgis/web/generic-parallel.html>.
 
 ## Bindings
 
diff --git a/docs/web/CMakeLists.txt b/docs/web/CMakeLists.txt
index 7667a75..ac56eb4 100644
--- a/docs/web/CMakeLists.txt
+++ b/docs/web/CMakeLists.txt
@@ -91,6 +91,7 @@ if(MGIS_HAVE_PANDOC)
   mgis_pandoc_generate_html_page(release-notes-3.1.1 "--toc" "--toc-depth=3")
   mgis_pandoc_generate_html_page(release-notes-3.2 "--toc" "--toc-depth=3")
   mgis_pandoc_generate_html_page(orthotropic-behaviours "--toc" "--toc-depth=3")
+  mgis_pandoc_generate_html_page(generic-parallel "--toc" "--toc-depth=3")
   mgis_pandoc_generate_html_page(behaviour-integration-failure-analysis "--toc" "--toc-depth=3")
   mgis_pandoc_generate_html_page(functions "--toc" "--toc-depth=3")
   mgis_pandoc_generate_html_page(evaluators "--toc" "--toc-depth=3")
diff --git a/docs/web/behaviour-integration-failure-analysis.md b/docs/web/behaviour-integration-failure-analysis.md
index fbea6bc..4de5fb6 100644
--- a/docs/web/behaviour-integration-failure-analysis.md
+++ b/docs/web/behaviour-integration-failure-analysis.md
@@ -52,6 +52,20 @@ const auto r = integrate_debug(v, b, a);
 
 See Section @sec:mgis:integration_failure:custom_analyser for details.
 
+## Behaviours generated by the `generic-parallel` interface
+
+The `integrate_debug` functions are not available for a behaviour
+generated by the `generic-parallel` interface of `MFront`: they throw an
+exception. Such a behaviour integrates all the points of a batch in one
+call, and only reports the index of the first point where the
+integration failed (member `n` of the `BehaviourIntegrationResult`
+structure, see [this page](generic-parallel.html)).
+
+The same `MFront` file, compiled with the `generic` interface, gives the
+detailed diagnostic at this point: a material data manager of this
+behaviour, on the host, is given the data of the point, and the point is
+integrated by the `integrate_debug` function taking a range of points.
+
 ## Description of the default analyser {#sec:mgis:integration_failure:default_analyser}
 
 The default analyser generates one file per integration failure
diff --git a/docs/web/bindings-cxx.md b/docs/web/bindings-cxx.md
index 4335040..fac3330 100644
--- a/docs/web/bindings-cxx.md
+++ b/docs/web/bindings-cxx.md
@@ -86,11 +86,22 @@ generated by `MFront`.
 > provided by the `TFEL` project. Those libraries must be available
 > in the current environment for the library to be usable in `MGIS`.
 
+> **Behaviours generated by the `generic-parallel` interface**
+>
+> `MFront` can also generate libraries which integrate batches of
+> integration points (`mfront --interface=generic-parallel`), on the
+> host or on a GPU, with a backend chosen at generation time. Those
+> libraries are loaded by the same `load` functions. They are also linked
+> to the runtime of their backend (`CUDA`, `HIP`, `TBB`, `Kokkos`,
+> `SYCL`, `OpenMP`). Their use is described in
+> [this page](generic-parallel.html).
+
 # Loading the behaviour, the `Behaviour` object
 
 The `load` function is used to load the behaviour. This function returns
 a `Behaviour` object which contains all the relevant information about
-the behaviour and a function pointer.
+the behaviour and a function pointer: the `b` member for the `generic`
+interface, the `batch_b` member for the `generic-parallel` interface.
 
 ## Small strain behaviours
 
@@ -184,8 +195,8 @@ It contains two data members:
 - the name
 - the type
 
-For material properties (and external state variables), only scalar
-variables are allowed.
+Material properties are scalars. External state variables may be
+scalars, vectors or tensors.
 
 With those informations, a solver developer is already able to check the
 validity of an input file.
@@ -219,6 +230,15 @@ Thus, the `BehaviourDataView` structure, and the underlying
 `StateView` structure, are now discussed. Then more details
 about the three cases described previously are given.
 
+> **Note**
+>
+> A behaviour generated by the `generic-parallel` interface has no
+> function integrating one integration point. Only the second case
+> applies: the functions acting on a `MaterialDataManager` integrate
+> all the points as a batch, described by a `BatchDataView` structure
+> which refers to the arrays of the manager. See
+> [this page](generic-parallel.html).
+
 ## The `BehaviourDataView` and `StateView` structures
 
 The `BehaviourDataView` data structure contains the following data
@@ -409,7 +429,10 @@ used to store all the data required.
 
 However, one may want to use memory allocated by the solver. In this
 case, one can have a look at the `MaterialDataManagerInitializer`
-class.
+class. This class also declares the memory location of the arrays (host,
+device or shared memory) and can give a memory resource used to allocate
+the arrays which are not given. See [this page](generic-parallel.html)
+for details.
 
 ### Updating the date from one time step to the other
 
@@ -423,6 +446,9 @@ The `revert` function overwrites the `s1` member, i.e. the state at the
 end of the time step, with `s0`, i.e. the state at the beginning of the
 time step.
 
+The `update` and `revert` functions copy the values on the host: they
+throw an exception if the arrays are located in the memory of a device.
+
 ## Third case: manual allocation
 
 The `MGIS` library provides a set of functions which allows the
@@ -443,7 +469,7 @@ strain, which is a scalar. Thus, the value of `n` will be equal to `7`.
 
 # Behaviour integration
 
-Three overloads of the `integrate` function are available:
+Three main overloads of the `integrate` function are available:
 
 The first one has the following interface:
 
@@ -464,11 +490,21 @@ argument of the `ThreadPool` type allowing a parallelization of the
 behaviour integration. The user may refer to the `IntegrateTest2b.cxx`
 file to have an example on how this use this second overload.
 
-> ** Note **
+For a behaviour generated by the `generic-parallel` interface, the
+overloads taking a `MaterialDataManager` integrate the points as one
+batch, with the execution policy of the manager (the thread pool is not
+used). Additional overloads take an execution policy
+(`mgis::BatchExecutionPolicy`) as first argument. The first overload
+throws an exception. See [this page](generic-parallel.html).
+
+> **Note**
 >
 > The `integrate_debug` function can be used to generate debug files
 > to analyse integration failures. See
 > [this page](behaviour-integration-failure-analysis.html) for details.
+> It is not available for the `generic-parallel` interface: the library
+> generated by the `generic` interface for the same behaviour shall be
+> used instead.
 
 # Support for behaviours' initialize functions {#sec:mgis:2.1:initialize_functions}
 
@@ -493,6 +529,14 @@ point or a set of integrations points.
 
 The `executeInitializeFunction` functions execute an initialization
 function on a unique integration point or a set of integration points.
+If the initialize function fails at an integration point of a set, the
+exit status of the result is `-1`.
+
+For a behaviour generated by the `generic-parallel` interface, the
+pointer to the initialize function of a batch is the `batch_f` member,
+and the function acting on a unique integration point throws an
+exception. The functions acting on a `MaterialDataManager` process all
+the points as a batch, as the `integrate` function.
 
 > **Note**
 >
@@ -525,10 +569,17 @@ set of integrations points.
 The `executePostProcessing` functions execute a postprocessing on a
 unique integration point or a set of integration points.
 
+For a behaviour generated by the `generic-parallel` interface, the
+pointer to the post-processing of a batch is the `batch_f` member, and
+the function acting on a unique integration point throws an exception.
+The outputs must be accessible by the computations of the behaviour:
+`allocatePostProcessingVariables` allocates them on the host.
+
 # Orthotropic behaviours
 
 Orthotropic behaviours require to handle the rotations between the
 global frame and the material frame. This is described in depth
-[here](orthotropic-behaviours.html).
+[here](orthotropic-behaviours.html), including the rotations of batches
+of integration points.
 
 # References
\ No newline at end of file
diff --git a/docs/web/generic-parallel.md b/docs/web/generic-parallel.md
new file mode 100644
index 0000000..7e76d6b
--- /dev/null
+++ b/docs/web/generic-parallel.md
@@ -0,0 +1,289 @@
+---
+title: Batches of integration points (`generic-parallel` interface)
+date: 2026
+lang: en-EN
+numbersections: true
+documentclass: article
+from: markdown+tex_math_single_backslash
+geometry:
+  - margin=2cm
+papersize: a4
+link-citations: true
+colorlinks: true
+figPrefixTemplate: "$$i$$"
+tabPrefixTemplate: "$$i$$"
+secPrefixTemplate: "$$i$$"
+eqnPrefixTemplate: "($$i$$)"
+---
+
+The `generic-parallel` interface of `MFront` generates, for a behaviour, a
+model or a material property, functions processing a whole batch of
+points in one call. The same `MFront` file can be compiled for several
+backends: `CUDA`, `HIP`, the parallel algorithms of the `C++` standard
+library, `Kokkos`, `SYCL` and `OpenMP`. A generated library targets one
+backend, and may compute on the host or on a GPU. The interface is
+described in the documentation of `MFront`.
+
+`MGIS` recognizes those libraries when they are loaded and uses them
+through its usual objects (`Behaviour`, `MaterialDataManager`,
+`MaterialProperty`). This page describes how.
+
+# Overview
+
+- The existing functions of `MGIS` acting on a material data manager
+  (`integrate`, `executeInitializeFunction`, `executePostProcessing`, the
+  rotations) automatically switch to the batch functions of a
+  `generic-parallel` library. The calling code is unchanged.
+- Each batch function also exists with an execution policy, which selects
+  the queue of the computations (`CUDA` stream, `SYCL` queue, etc.).
+- No data is copied: the batch is described by views on the arrays of the
+  material data manager.
+- The calling code declares where its arrays are located. `MGIS` checks
+  that the computations of the library can access them.
+- `MGIS` does not depend on any GPU toolkit. The arrays are given by the
+  calling code, or allocated by a memory resource given by the calling
+  code.
+
+# Loading
+
+The `load` functions read the symbols describing a library. For a library
+generated by the `generic-parallel` interface, the version of its data
+structures is checked, and the backend and the memory space of the library
+are read. They are stored in the description of the behaviour (members
+`interface`, `backend` and `memory_space`). The
+`isGenericParallelBehaviour` function tells if a behaviour processes
+batches.
+
+`MGIS` has its own copies of the data structures of `TFEL`
+(`MGIS/Behaviour/BatchDataView.hxx`), as for the `generic` interface: it
+does not depend on `TFEL` at compile time.
+
+The functions acting on a single integration point (`integrate` on a
+`BehaviourDataView`, `integrate_debug`, the initialize functions, the
+post-processings and the rotations of a point, the `C` bindings) report an
+error for a `generic-parallel` library. Conversely, the batch functions
+report an error for a library of the `generic` interface.
+
+# Integration of a batch
+
+~~~~{.cxx}
+auto r = integrate(policy, m, opts, dt);        // all the points
+auto r = integrate(policy, m, opts, dt, b, e);  // points [b, e)
+~~~~
+
+The overloads without execution policy use the policy of the material
+data manager (`batch_execution_policy` member). The overloads taking a
+thread pool make a single call: the parallelism is the one of the
+backend.
+
+The data of the material data manager are mapped as follows:
+
+- the gradients, the thermodynamic forces, the internal state variables
+  and the energies are fields stored point by point;
+- the material properties and the external state variables are given
+  variable by variable: each one is uniform or given point by point, as in
+  the `MaterialStateManager`. The elements of an array of variables are
+  distinct variables;
+- the mass density is only given when the speed of sound is computed.
+
+Contrary to the integration point by point, all the points of the batch
+are integrated, even if the integration fails at some points. The result
+gives the index of the first point where the integration failed
+(member `n`, an absolute index) and the number of failures in the error
+message. The detailed error message of a point can be obtained by
+integrating this point with the library of the `generic` interface of the
+same behaviour (see [this page](behaviour-integration-failure-analysis.html)).
+
+# Initialize functions, post-processings and rotations
+
+They follow the same pattern:
+
+~~~~{.cxx}
+executeInitializeFunction(policy, m, name[, inputs][, b, e]);
+executePostProcessing(outputs, policy, m, name[, b, e]);
+rotateGradients(policy, destination, behaviour, source, r);
+~~~~
+
+The overloads without execution policy switch automatically. The
+rotation matrices can be uniform or given point by point. The
+`RotationMatrix2D` and `RotationMatrix3D` objects build the rotation
+matrices on the host at each call: for repeated rotations on a GPU, the
+matrices shall rather be built once and given as an array.
+
+# Models
+
+The models generated by the `generic-parallel` interface export the same
+functions and symbols as the behaviours. `mgis::model::load` loads them as
+behaviours: they are integrated by the same batch functions, and their
+parameters are set by `setParameter`.
+
+# Material properties
+
+~~~~{.cxx}
+const auto mp = mgis::material_property::load(library, "YoungModulus");
+const auto inputs = std::array{std::span<const mgis::real>(temperatures)};
+const auto s1 = evaluate(values, mp, inputs);
+const auto s2 = evaluate(policy, values, mp, inputs, MemoryLocation::SHARED);
+~~~~
+
+When a material property generated by the `generic-parallel` interface is
+loaded, `MGIS` reads the version of its functions, its backend and its
+memory space, and loads the function evaluating a set of points
+(`batch_fct` member) instead of the function evaluating one point (`fct`
+member). `isGenericParallelMaterialProperty` tells if a material property
+evaluates sets of points.
+
+The `evaluate` function evaluates a material property on a set of points.
+Each input is uniform (one value) or given point by point. As
+`integrate`, it makes one call for a `generic-parallel` material property,
+and evaluates the other material properties point by point on the host.
+The returned status follows the conventions of the `generic` interface of
+the material properties. For a `generic-parallel` material property, it
+summarizes all the points: no index of point is reported.
+
+# Memory
+
+## Location of the arrays and memory space of a library
+
+The calling code declares the location of its arrays
+(`mgis::MemoryLocation`):
+
+- `HOST`: ordinary memory of the host;
+- `DEVICE`: memory of a device, not accessible by the host;
+- `SHARED`: memory accessible by the host and the device (managed memory
+  of `CUDA` or `HIP`, shared memory of `SYCL`, `Kokkos::SharedSpace`).
+
+The location of the arrays of a material data manager is given by its
+initializer (`location` member of `MaterialDataManagerInitializer`). The
+location of the values of a material property is an argument of
+`evaluate`.
+
+`MGIS` compares this location to the memory space exported by the
+library (`isMemoryLocationAccessible`):
+
+| Memory space of the library | `HOST` | `DEVICE` | `SHARED` |
+|---|---|---|---|
+| `host` (and libraries of the `generic` interface) | yes | no | yes |
+| `device` | no | yes | yes |
+| `any` | yes | yes | yes |
+
+`MGIS` thus knows nothing about the backends: a new backend of `TFEL`
+requires no change in `MGIS`.
+
+## Functions working on the host
+
+Some functions of `MGIS` read the arrays on the host: `update`, `revert`,
+`updateValues`, `extractInternalStateVariable`, `save`, `restore` and the
+integration point by point. They report an error if the arrays are located
+on a device. `allocatePostProcessingVariables` allocates the outputs of a
+post-processing on the host: if the behaviour can't access them, they
+must be allocated by the calling code.
+
+The values held by `MGIS` (uniform values and values stored with the
+`LOCAL_STORAGE` mode) are stored on the host. They are refused when the
+computations can't access the memory of the host: they must be given by a
+`std::span` on accessible memory, with the `EXTERNAL_STORAGE` mode. A
+uniform value can be given by a `std::span` of one value.
+
+## Memory resources
+
+A material data manager allocates the arrays that its initializer does not
+give. By default, those arrays are allocated on the host. The initializer
+can give a memory resource (`resource` member, a
+`std::pmr::memory_resource*`), used to allocate them. `MGIS` initializes
+those arrays on the host: the resource must allocate memory accessible by
+the host, in practice shared memory. If the arrays are located on a
+device, all of them must be given by the initializer.
+
+Four memory resources are provided as headers, compiled by the calling
+code with its own GPU toolkit (`MGIS/MemoryResources/`):
+
+| Header | Memory |
+|---|---|
+| `CUDAManagedMemoryResource.hxx` | managed memory of `CUDA` |
+| `HIPManagedMemoryResource.hxx` | managed memory of `HIP` |
+| `SYCLSharedMemoryResource.hxx` | shared memory of `SYCL`, for a given queue |
+| `KokkosSharedSpaceMemoryResource.hxx` | `Kokkos::SharedSpace` |
+
+# Execution policy
+
+`mgis::BatchExecutionPolicy` is shared by the behaviours, the models and
+the material properties:
+
+- `queue`: queue of the computations, whose type depends on the backend
+  (`cudaStream_t`, `hipStream_t`, pointer to a `sycl::queue` or to an
+  instance of a `Kokkos` execution space). A null pointer denotes the
+  default queue, if meaningful for the backend (a queue is required by the
+  `SYCL` backend).
+- `preserve_stack_size`: if true, the stack size of the threads of the
+  device is never modified, and an error is reported if it is too small.
+  Otherwise, the stack size is increased if required. This affects the
+  whole process, since the stack size is a property of the context of the
+  device (`CUDA` and `HIP`).
+
+The overloads without execution policy use the policy of the material data
+manager, or a default policy for the rotations and the material
+properties. The default policy gives a null queue, which the `SYCL`
+backend refuses.
+
+The computations performed by `MGIS` are synchronous: the functions return
+once the computations are completed.
+
+# Errors
+
+`MGIS` distinguishes two cases:
+
+- a failure of the computations at some points (lack of convergence,
+  physical bound violated) is a result: the status is `-1`, and the index
+  of the first point where the computations failed is reported;
+- invalid arguments and errors of the backend are reported by exceptions.
+  They are never reported as a failure of the computations.
+
+For the material properties, the statuses `-5`, `-6` and `-7` (invalid
+arguments, runtime error of the backend, invalid strides) are reported
+by exceptions. As with the `generic` interface, the status `-2` (exception
+thrown on the host by the generated function) is returned.
+
+The batch functions report errors by exceptions, as the other functions
+acting on a material data manager. They have no variant taking an
+execution context.
+
+# Current limitations
+
+- The asynchronous computations offered by `TFEL` are not exposed.
+- The optional arrays giving the status and the time step scaling factor
+  of each point are not exposed.
+- The `C`, `Fortran` and `Python` bindings do not expose the new
+  functions. The `C` and `Fortran` functions acting on a single point
+  report a failure for a `generic-parallel` library.
+
+# Example
+
+A behaviour compiled for `CUDA`, the arrays of the material data manager
+allocated by `MGIS` in managed memory, the computations submitted to a
+stream chosen by the calling code. The temperature is given by a
+`std::span` of one value located in managed memory.
+
+~~~~{.cxx}
+#include "MGIS/Behaviour/Behaviour.hxx"
+#include "MGIS/Behaviour/MaterialDataManager.hxx"
+#include "MGIS/Behaviour/Integrate.hxx"
+#include "MGIS/MemoryResources/CUDAManagedMemoryResource.hxx"
+
+using namespace mgis::behaviour;
+// the resource must outlive the material data manager
+auto resource = mgis::memory_resources::CUDAManagedMemoryResource{};
+const auto b =
+    load("libBehaviour-cuda.so", "Norton", Hypothesis::TRIDIMENSIONAL);
+auto i = MaterialDataManagerInitializer{};
+i.location = MemoryLocation::SHARED;
+i.resource = &resource;
+auto m = MaterialDataManager(b, n, i);
+for (auto* const s : {&m.s0, &m.s1}) {
+  setExternalStateVariable(*s, "Temperature", temperature,
+                           MaterialStateManager::EXTERNAL_STORAGE);
+}
+auto policy = BatchExecutionPolicy{};
+policy.queue = stream;
+const auto r = integrate(policy, m, BehaviourIntegrationOptions{}, dt);
+~~~~
diff --git a/docs/web/index.md b/docs/web/index.md
index 2dc497a..2ae353d 100644
--- a/docs/web/index.md
+++ b/docs/web/index.md
@@ -28,13 +28,14 @@ bibliography: bibliography.bib
 
 ![Principle of MGIS](img/mgis.svg){width=100%}
 
-This project aims at proving tools (functions, classes, bindings,
-etc...) to handle behaviours written using `MFront` generic interface.
+This project aims at providing tools (functions, classes, bindings,
+etc...) to handle behaviours, models and material properties generated by
+the `generic` and `generic-parallel` interfaces of `MFront`.
 For information about `MFront`, see
 <http://thelfer.github.io/tfel/web/index.html>.
 
 Those tools are meant to be used by (`FEM`, `FFT`, etc.) solver
-developers. This tools are *not* linked to the `TFEL` libraries.
+developers. Those tools are *not* linked to the `TFEL` libraries.
 Permissive licences have been chosen to allow integration in open-source
 and proprietary codes.
 
@@ -54,7 +55,15 @@ library provides two main functions:
   step.
 
 The library also supports handling a group of integration points though
-the `mgis::behaviour::MaterialStateManager` class.
+the `mgis::behaviour::MaterialStateManager` and
+`mgis::behaviour::MaterialDataManager` classes.
+
+The behaviours, models and material properties generated by the
+`generic-parallel` interface of `MFront` process whole batches of
+integration points, on CPUs or on GPUs. The functions acting on a
+`MaterialDataManager` use them automatically. Material properties are
+evaluated on a set of points by the `mgis::material_property::evaluate`
+function. Their use by `MGIS` is described [here](generic-parallel.html).
 
 An introduction to the `C++` library may be found [here](bindings-cxx.html)
 
diff --git a/docs/web/install.md b/docs/web/install.md
index 09136ee..9323c23 100644
--- a/docs/web/install.md
+++ b/docs/web/install.md
@@ -70,7 +70,7 @@ The simplest way to compile this package is:
 - `enable-fortran-bindings`: compiles bindings for the `Fortran2003`
   language (default=OFF)
 - `enable-python-bindings`: compiles the `Python` bindings. This requires
-  the `Boost/Python` library to be available (default=OFF)
+  the `pybind11` library to be available (default=OFF)
 - `enable-fenics-bindings`: compiles the `FEniCS` bindings. Those
   bindings are experimental and very limited. To use `MGIS` with
   `FEniCS`, usage of the `Python` bindings are encouraged.
@@ -82,6 +82,16 @@ The simplest way to compile this package is:
 - `enable-static`: compiles static libraries
 - `enable-doxygen-doc`: enable the generation of the API documentation
   using with `Doxygen`.
+- `enable-generic-parallel-cuda-tests`: test the behaviours, models and
+  material properties generated by the `generic-parallel` interface of
+  `MFront` with the `cuda` backend. This requires a `CUDA` compiler and an
+  `NVIDIA` GPU (default=OFF).
+
+The tests of the `generic-parallel` interface of `MFront` are only built
+if the `TFEL` libraries found provide this interface. They use the
+`stlpar` backend (parallel algorithms of the `C++` standard library). With
+`gcc`, those algorithms rely on the `TBB` library, which is linked if
+found.
 
 By default, the library tries to autodetect the TFEL libraries and
 expects that a working version of TFEL is available in the
diff --git a/docs/web/mgis-template.html b/docs/web/mgis-template.html
index d27d1f8..36c8022 100644
--- a/docs/web/mgis-template.html
+++ b/docs/web/mgis-template.html
@@ -143,6 +143,7 @@
 	      <ul>
 		<li><a href="orthotropic-behaviours.html">Support for orthotropic behaviours</a></li>
 		<li><a href="behaviour-integration-failure-analysis.html">Analysing failure of behaviour integration</a></li>
+		<li><a href="generic-parallel.html">Batches of integration points (generic-parallel interface)</a></li>
 		<li><a href="functions.html">Functions</a></li>
 	      </ul>
 	    </li>
diff --git a/docs/web/orthotropic-behaviours.md b/docs/web/orthotropic-behaviours.md
index b194d53..919ea9a 100644
--- a/docs/web/orthotropic-behaviours.md
+++ b/docs/web/orthotropic-behaviours.md
@@ -26,8 +26,10 @@ behaviours.
 However, dealing with orthotropic behaviours may be tedious and require
 an considerable amount of developments. Since version \(3.4\) of `TFEL`,
 the `generic` interface generates additional functions which performs
-those steps efficiently. The purpose of this document is to introduce
-some of the features provided by `MGIS` built around those functions.
+those steps efficiently. The `generic-parallel` interface generates the
+same functions for batches of integration points. The purpose of this
+document is to introduce some of the features provided by `MGIS` built
+around those functions.
 
 ## Outline of this document
 
@@ -36,7 +38,9 @@ introduction on the specific issues related to orthotropic behaviours
 from the behaviour integration point of view.
 
 The functions generated by the `MFront`' `generic` interface are
-described in Section @sec:mgis:orthotropic_behaviours:generic_interface.
+described in Section @sec:mgis:orthotropic_behaviours:generic_interface,
+and the ones generated by the `generic-parallel` interface in Section
+@sec:mgis:orthotropic_behaviours:generic_parallel_interface.
 
 Section @sec:mgis:orthotropic_behaviours:mgis_support is devoted to
 functionalities and wrappers around those functions and how to use them
@@ -101,7 +105,7 @@ functions are also generated:
 Those functions takes an additional arguments which is the number of
 integration points to be treated.
 
-### Finite strain behaviours
+## Finite strain behaviours
 
 Finite strain behaviours are a special case, because the returned stress
 measure and the returned tangent operator can be chosen at runtime time.
@@ -122,6 +126,7 @@ Here is the list of the generated functions:
   `<behaviour_function_name>_<hypothesis>_rotateThermodynamicForces_PK2Stress`.
   This function assumes that its first argument is the second
   Piola-Kirchhoff stress in the material frame.
+-
   `<behaviour_function_name>_<hypothesis>_rotateTangentOperatorBlocks_dsig_dF`.
   This function assumes that its first argument is the derivative of the
   Cauchy stress with respect to the deformation gradient in the material
@@ -132,10 +137,46 @@ Here is the list of the generated functions:
   first Piola-Kirchhoff stress with respect to the deformation gradient
   in the material frame.
 -
-  `<behaviour_function_name>_<hypothesis>_rotateTangentOperatorBlocks_PK2Stress`.
+  `<behaviour_function_name>_<hypothesis>_rotateTangentOperatorBlocks_dPK2_dEGL`.
   This function assumes that its first argument is the derivative of the
   second Piola-Kirchhoff stress with respect to the Green-Lagrange
   strain in the material frame.
+-
+  `<behaviour_function_name>_<hypothesis>_rotateTangentOperatorBlocks_dtau_ddF`.
+  This function assumes that its first argument is the derivative of the
+  Kirchhoff stress with respect to the spatial increment of the
+  deformation gradient in the material frame.
+
+The functions treating arrays of integration points follow the same
+convention, for instance
+`<behaviour_function_name>_<hypothesis>_rotateArrayOfThermodynamicForces_CauchyStress`.
+
+# Functions generated by the `generic-parallel` interface {#sec:mgis:orthotropic_behaviours:generic_parallel_interface}
+
+The `generic-parallel` interface generates the same functions for
+batches of integration points:
+
+- `<behaviour_function_name>_<hypothesis>_rotateGradients`
+- `<behaviour_function_name>_<hypothesis>_rotateThermodynamicForces`
+- `<behaviour_function_name>_<hypothesis>_rotateTangentOperatorBlocks`
+
+and, for finite strain behaviours, the same variants for each stress
+measure and each tangent operator. There are no `rotateArrayOf` variants:
+all the functions treat a batch of points.
+
+All those functions have the following signature:
+
+~~~~{.cxx}
+int f(mgis_bv_BatchOutputStatus*, const mgis_bv_BatchRotationDataView*);
+~~~~
+
+The `mgis_bv_BatchRotationDataView` structure gives the number of points,
+the rotated values, the values to be rotated, which may be the rotated
+values, and the rotation matrices from the global frame to the material
+frame. Each rotation matrix has 9 components stored in column-major
+format, as for the `generic` interface. The rotation matrix is uniform if
+the point stride of its field is null. The computations are submitted to
+the queue given by the execution options of the structure.
 
 # Features of `MGIS` related to orthotropic behaviours {#sec:mgis:orthotropic_behaviours:mgis_support}
 
@@ -155,6 +196,14 @@ modelling hypothesis and, for finite strain behaviours, the choices of
 the stress measures and tangent operator kind, are automatically taken
 into account by the `load` functions.
 
+For a behaviour generated by the `generic-parallel` interface, those
+pointers are null. The rotation functions of batches are given by the
+following members:
+
+- `batch_rotate_gradients_ptr`
+- `batch_rotate_thermodynamic_forces_ptr`
+- `batch_rotate_tangent_operator_blocks_ptr`
+
 Direct use of those functions is meant to be highly efficient but still
 don't perform any consistency checks.
 
@@ -166,8 +215,21 @@ The following helper functions are thus provided:
 
 Those functions takes memory blocks (wrapped in `std::span` objects)
 and checks that the size of those blocks are consistent. For each of
-those functions, two overloads exist, one for in place rotation, one for
-out of place rotations.
+those functions, overloads exist for in place and for out of place
+rotations. The rotation matrix is given either by an array (9 values for
+a uniform rotation, or 9 values per integration point) or by a
+`RotationMatrix2D` or a `RotationMatrix3D` object, which builds the
+rotation matrices from the material axes.
+
+For a behaviour generated by the `generic-parallel` interface, those
+functions rotate all the points by one call of the batch rotation
+function. Additional overloads take an execution policy
+(`mgis::BatchExecutionPolicy`) as first argument, which selects the queue
+of the computations. The arrays must be accessible by the computations of
+the behaviour. The matrices of the `RotationMatrix2D` and
+`RotationMatrix3D` objects are built on the host at each call: for
+repeated rotations on a GPU, the matrices shall be built once and given as
+an array (see [this page](generic-parallel.html)).
 
 Those functions are perfectly fine for managing the rotations of the
 gradients of a set of integration points. However, their overhead is
@@ -209,6 +271,9 @@ The `C` bindings provides the following functions, declared in the
 - `mgis_bv_behaviour_rotate_array_of_tangent_operator_blocks_in_place`
 - `mgis_bv_behaviour_rotate_array_of_tangent_operator_blocks_out_of_place`
 
+Those functions report a failure for a behaviour generated by the
+`generic-parallel` interface.
+
 ## `Fortran` bindings
 
 The `mgis.behaviours` bindings provides the following functions, declared in the
@@ -227,6 +292,9 @@ The `mgis.behaviours` bindings provides the following functions, declared in the
 - `rotate_array_of_tangent_operator_blocks_in_place`
 - `rotate_array_of_tangent_operator_blocks_out_of_place`
 
+As the `C` functions, they report a failure for a behaviour generated by
+the `generic-parallel` interface.
+
 ### Example of usage
 
 ~~~~{.fortran}
diff --git a/docs/web/release-notes-3.2.md b/docs/web/release-notes-3.2.md
index 3839baf..15ccebe 100644
--- a/docs/web/release-notes-3.2.md
+++ b/docs/web/release-notes-3.2.md
@@ -25,6 +25,23 @@ This version is meant to be used with `TFEL` Version 5.2.
 - Modifiers in `MGIS/Function` now creates a view from a temporary view
   rather than an evaluator. See below for details. In practice, most
   views are also evaluators, so this shall not impact existing user code.
+- A uniform material property given by a `std::span` of one value with
+  the `LOCAL_STORAGE` mode is now stored as a scalar, as the external state
+  variables: `isMaterialPropertyUniform` returns `true` for it.
+- The `fct` member of a material property generated by the
+  `generic-parallel` interface of `MFront` is null: such a material
+  property is evaluated by the `evaluate` function (see below).
+- The functions of the `LibrariesManager` class returning the functions
+  of a point (`getBehaviour`, `getMaterialProperty`,
+  `getBehaviourInitializeFunction`, `getBehaviourPostProcessing`, the
+  rotation functions) throw an exception for an entry point generated by
+  the `generic-parallel` interface.
+- The failure of an initialize function or of a post-processing executed
+  on a material data manager now sets the exit status of the result to
+  `-1`. It was previously left to `1`.
+- The arrays held by the material state managers and the material data
+  managers are now `std::pmr::vector` objects. The `API` is unchanged, but
+  the calling codes must be recompiled.
 
 # Documentation
 
@@ -35,6 +52,13 @@ This version is meant to be used with `TFEL` Version 5.2.
 Support for `CUDA` in `MGIS/Function` is enable by
 `enable-cuda-algorithms`.
 
+### Tests of the `generic-parallel` interface with `CUDA`
+
+The `enable-generic-parallel-cuda-tests` option enables the tests of the
+behaviours, models and material properties generated by the
+`generic-parallel` interface of `MFront` with the `cuda` backend. Those
+tests require a `NVIDIA` GPU. `OFF` by default.
+
 ### HDF5 support
 
 The following options control the support of the `HDF5` library:
@@ -95,6 +119,40 @@ directory (refered to `<install_prefix>` in the following):
 > If `MGIS` is built with `TFEL` support, the `TFEL` environment
 > shall be properly set.
 
+# Support of the `generic-parallel` interface of `MFront`
+
+The behaviours, models and material properties generated by the
+`generic-parallel` interface of `MFront` process whole batches of points,
+on CPUs or on GPUs (`CUDA`, `HIP`, parallel algorithms of the `C++`
+standard library, `Kokkos`, `SYCL`, `OpenMP`). Their support is described
+in a dedicated [page](generic-parallel.html). This support requires a
+version of `TFEL` providing those interfaces (version 2 of the data
+structures of the behaviours and of the functions of the material
+properties). In short:
+
+- the libraries are recognized when they are loaded, and their version,
+  backend and memory space are checked;
+- the existing functions acting on a material data manager
+  (`integrate`, `executeInitializeFunction`, `executePostProcessing`, the
+  rotations) automatically switch to the batch functions, without copying
+  any data;
+- each batch function also takes an execution policy
+  (`mgis::BatchExecutionPolicy`), which selects the queue of the
+  computations;
+- the location of the arrays (`mgis::MemoryLocation`: host, device or
+  shared memory) is declared by the calling code and checked against the
+  memory space of the library. The functions working on the host refuse
+  arrays located on a device;
+- a material data manager can allocate its arrays with a memory resource
+  given by its initializer. Memory resources for the managed memory of
+  `CUDA` and `HIP`, the shared memory of `SYCL` and `Kokkos` are provided
+  as headers (`MGIS/MemoryResources`);
+- the `evaluate` function evaluates a material property on a set of
+  points, as a batch for a `generic-parallel` material property, point by
+  point otherwise;
+- the functions acting on a single point report an error for a
+  `generic-parallel` library instead of calling a null pointer.
+
 # New features of the `MGIS/Behaviour` library
 
 ## Save/restore operations in `MaterialStateManager` and `MaterialDataManager`
@@ -337,6 +395,19 @@ For more details, see <https://github.com/thelfer/MFrontGenericInterfaceSupport/
 
 For more details, see <https://github.com/thelfer/MFrontGenericInterfaceSupport/issues/158>￼
 
+## Other fixes
+
+- The checks of the material axes of the `RotationMatrix2D` and
+  `RotationMatrix3D` objects were inverted, and the number of points was
+  compared to a remainder instead of a quotient: no such object could be
+  built.
+- A material property can be given by a `std::span` of one value, as an
+  uniform value.
+- The failure of an initialize function or of a post-processing at an
+  integration point now sets the exit status to `-1` (see the known
+  incompatibilities).
+- The mass density can be given by a `std::span` of one value.
+
 # Acknowledgements
 
 The authors are grateful to the many contributors to the `TFEL/MFront`
-- 
2.43.0

